p2p/peer_set.go

```diff
@@ -47,6 +47,9 @@ func (ps *PeerSet) Add(peer Peer) error {
 	if ps.lookup[peer.ID()] != nil {
 		return ErrSwitchDuplicatePeerID{peer.ID()}
 	}
+	if peer.GetRemovalFailed() {
+		return ErrPeerRemoval{}
+	}
```

Comment:
Can we do this check before calling `PeerSet.Add` here. I would prefer to do it on line 813 of switch.go before initializing the peer to the individual reactors

Reply:
Will this not change the effectiveness of the fix?

Reply:
What do you mean by effectiveness? As in it will no longer fix it?

Reply:
If this check happens anywhere earlier we introduce the exact same race condition we already heaving.

Reply:
> If this check happens anywhere earlier we introduce the exact same race condition we already heaving.

I see. I wasn't aware of that. Need to look in more closely then